app/demo-interactor/KNDemoKernel.cu

```diff
@@ -0,0 +1,191 @@
+//---------------------------------*-CUDA-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file KNDemoKernel.cu
+//---------------------------------------------------------------------------//
+#include "KNDemoKernel.hh"
+
+#include <thrust/device_ptr.h>
+#include <thrust/reduce.h>
+#include "base/Assert.hh"
+#include "physics/base/ParticleTrackView.hh"
+#include "physics/base/SecondaryAllocatorView.hh"
+#include "physics/em/KleinNishinaInteractor.hh"
+#include "random/cuda/RngEngine.cuh"
+
+using namespace celeritas;
+
+namespace demo_interactor
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Kernel to initialize particle data.
+ *
+ * For testing purposes (this might not be the case for the final app) we use a
+ * grid-stride loop rather than requiring that each thread correspond exactly
+ * to a particle track. In other words, this method allows a single warp to
+ * operate on two 32-thread chunks of data.
+ *  https://developer.nvidia.com/blog/cuda-pro-tip-write-flexible-kernels-grid-stride-loops/
+ */
+__global__ void initialize_kn(const ParticleParamsPointers params,
+                              const ParticleStatePointers  states,
+                              const ParticleTrackState     initial_state,
+                              Real3*                       direction,
+                              bool*                        alive)
+{
```

Comment:
It does not matter for this test, however it does looks like cudaMemcpy would/could work fine to do those intialization.

Reply:
That is an interesting point, but I imagine it's always going to be faster to do memory initialization directly in a device kernel, rather than allocating new memory on CPU, filling it up serially, and then copying it over. Is there something I'm missing? Your comment strikes me as similar to the philosophy assumed by VecGeom where initialization happens on GPU.

Reply:
> rather than allocating new memory on CPU, filling it up serially, and then copying it ove

In this case the presumably 'expensive' one is still done (in the current code) on the CPU (i.e. ParticleTrackState) [and if it is done on the CPU it can be done in parallel in either case].
Also I would gather than the allocation are done rarely (even on the CPU) and the memory re-used rather than free/re-allocated.

> Your comment strikes me as similar to the philosophy assumed by VecGeom where initialization happens on GPU.

Part of the initialization has no choice but to be on the CPU as the information comes from either a file or an external library (eg. Pythia).

Reply:
I don't understand your first comment. But the second one does make a lot of sense -- yes there's going to have to be a CPU copy in there somewhere. However, in general I think we'll be keeping the "initialization" data in a separate format from the "runtime" data -- CPU data will be more likely "array of structs" (energy, position, direction all packaged together) with GPU as "structs of array" (physics, geometry, etc. packaged independently).

Reply:
for the first point, I was partially off base (I forgot ParticleTrackState was small).  The remaining part/point is that " rather than allocating new memory on CPU, filling it up serially" does not have to be the case neither for the "allocating new memory" nor the "serially".

One thing that I/we can not tell from this exercise is what the real use case would look like because, **if I read the code correctly**, this kernel (and that's okay since this is a test) is placing the same particle type at the same place in all the slots, so the "hard part" of the initialization is not tested/tackled here (that's okay for this PR's purpose but I was thinking ahead).

Reply:
Gotcha. I'd assumed your comment was about the equivalent (simplistic, demo) kernel code.  You're absolutely correct that this isn't the hard part of real initialization, not by a long shot. I don't think the demo app even returns the time required for initialization... 😬